README.md

[![Arduino CI](https://github.com/RobTillaart/RunningMedian/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/RunningMedian/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/RunningMedian/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/RunningMedian/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/RunningMedian/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/RunningMedian/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/RunningMedian.svg?maxAge=3600)](https://github.com/RobTillaart/RunningMedian/releases)


# RunningMedian

Arduino library to determine the running median by means of a circular buffer.


## Description

Running Median looks like a running average with a small but important twist.
Running average averages the last N samples while the running median takes 
the last N samples, sort them and take the middle one, or the average of the 
middle two in case the internal buffer size is even.

Important differences between running average and running median:
- Running median will return real data (e.g. a real sample from a sensor) 
if one uses an odd size of the buffer (therefore preferred).
Running average may return a value that is never sampled.
- Running median will give zero weight to outliers, and 100% to the middle sample, 
whereas running average gives the same weight to all samples.
- Running median will give often constant values for some time.
- As one knows the values in the buffer one can predict the maximum change of 
the running median in the next steps in advance.
- Running median is slower as one needs to keep the values in timed order 
to remove the oldest and keep them sorted to be able to select the median.


#### Note: MEDIAN_MAX_SIZE

The maximum size of the internal buffer is defined by **MEDIAN_MAX_SIZE** and is 
set to 255 (since version 0.3.1). The memory allocated currently is in the order
of 5 bytes per element plus some overhead, so 255 elements take ~1300 bytes.
For an UNO this is quite a bit.

With larger sizes the performance penalty to keep the internal array sorted 
is large. For most applications a value much lower e.g. 19 is working well, and 
is performance wise O(100x) faster in sorting than 255 elements.


#### Note: Configurable Options

There are several options that can be configured via defines at compile time, those being:
- **RUNNING_MEDIAN_USE_MALLOC**: bool
  - true (default): Dynamic memory allocation is used for the buffer.
  - false: Static buffers of size MEDIAN_MAX_SIZE are used.
- **MEDIAN_MIN_SIZE**: uint8_t
  - Dynamic / Static: The buffer stores at least this many items.
  - should be minimal 3.
- **MEDIAN_MAX_SIZE**: uint8_t
  - Dynamic: Not used.
  - Static: The buffer stores at most this many items.


#### Related

- https://github.com/RobTillaart/Correlation
- https://github.com/RobTillaart/GST - Golden standard test metrics
- https://github.com/RobTillaart/Histogram
- https://github.com/RobTillaart/RunningAngle
- https://github.com/RobTillaart/RunningAverage
- https://github.com/RobTillaart/RunningMedian
- https://github.com/RobTillaart/statHelpers - combinations & permutations
- https://github.com/RobTillaart/Statistic


## Interface


### Constructor

- **RunningMedian(const uint8_t size)** Constructor, dynamically allocates memory.
- **~RunningMedian()** Destructor.
- **uint8_t getSize()** returns size of internal array.
- **uint8_t getCount()** returns current used elements, getCount() <= getSize().
- **bool isFull()** returns true if the internal buffer is 100% filled.


### Base functions

- **clear()** resets internal buffer and variables, effectively empty the buffer.
- **add(const float value)** adds a new value to internal buffer, 
optionally replacing the oldest element if the buffer is full.
- **float getMedian()** returns the median == middle element.
- **float getAverage()** returns average of **all** the values in the internal buffer.
- **float getAverage(uint8_t nMedian)** returns average of **the middle n** values. 
This effectively removes noise from the outliers in the samples.
The function is improved in 0.3.8 to correct a bias, see #22.
- **float getMedianAverage(uint8_t nMedian)** same as above, 
except it compensates for alignment bias, see #22.
This is done by adjusting the nMedian parameter (-1 or +1) if needed.
- **float getHighest()** get the largest values in the buffer.
- **float getLowest()** get the smallest value in the buffer.
- **float getQuantile(const float quantile)** returns the Quantile value from the buffer. 
This value is often interpolated.


### Less used functions

- **float getElement(const uint8_t n)** returns the n'th element from the values in time order.
- **float getSortedElement(const uint8_t n)** returns the n'th element from the values in size order (sorted ascending).
- **float predict(const uint8_t n)** predict the maximum change of median after n additions, 
n must be smaller than **getSize()/2**.


### SearchMode optimization

Since 0.3.7 the internal sort has been optimized.
It is now possible to select between LINEAR (=0) and BINARY (=1) insertion sort.
Pre-0.3.7 used linear insertion sort, and the new linear version is slightly optimized.
For larger internal arrays the performance gain of BINARY mode is substantial.

- **void setSearchMode(uint8_t searchMode = 0)** 0 = linear, 1 = binary - see table below.
Other values will set the searchMode to linear.
- **uint8_t getSearchMode()** returns the set mode

|  searchMode  |  value  | notes  |
|:------------:|:-------:|:-------|
|    LINEAR    |    0    |  fastest for smaller internal buffers (default)
|    BINARY    |    1    |  faster for larger internal buffers

Depends on the board / clock used where the methods are equally fast.

Give it a try, and let me know your.


## Operation

See examples.


## Future

- improve documentation.
- check for optimizations.
  - get the median without (full) sorting. QuickSelect()
- move all code to .cpp file




CHANGELOG.md
# Change Log RunningMedian

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.3.8] - 2023-07-11
- improved **getAverage(nMedians)** by removing bias - #22
  - thanks to Peter Kowald
- extended performance test
- update readme.md


## [0.3.7] - 2022-10-28
- Add RP2040 support to build-CI.
- Add CHANGELOG.md
- optimized binary insertion sort - see issue #20 (idea thanks to acicuc)
- add performance test sketch (to test sort efficiency)
- add setSearchMode() for selecting fastest median search mode.  
  EXPERIMENTAL, select between LINEAR or BINARY search.  

## [0.3.6} - 2022-06-06
- bump version for platformio

## [0.3.5} - 2022-06-05
- configuration options, 
- fixed static version not working

## [0.3.4} - 2021-12-28
- update library.json, 
- update readme.md, 
- license, minor edits

## [0.3.3} - 2021-01-22  
- better insertionSort (+ clean up test code)

## [0.3.2} - 2021-01-21  
- replaced bubbleSort by insertionSort
  better performance for large arrays.

## [0.3.1} - 2021-01-16
- Changed size parameter to 255 max

## [0.3.0} - 2021-01-04
- malloc() memory as default storage

----

## [0.2.2] - 2021-01-03
- add Arduino-CI + unit tests

## [0.2.1] - 2020-06-19
- fix library.json

## [0.2.0] - 2020-04-16
- refactor.

----

## [0.1.15] - 2018-08-24
- make runningMedian Configurable #110

## [0.1.14] - 2017-07-26
- revert double to float - issue #33

## [0.1.13] - 2015-10-30
- fix getElement(n) - kudos to Gdunge

## [0.1.12] - 2015-07-12
- refactor constructor + const

## [0.1.11] - 2015-03-29
- undo 0.1.10 fix clear

## [0.1.10] - 2015-03-07
- fix clear

## [0.1.09] - 2014-11-25
- float to double (support ARM)

## [0.1.08] - 2013-10-20
- add getElement(), 
- add getSottedElement() 
- add predict()

## [0.1.07] - 2013-10-19
- add correct median if \_count is even.

## [0.1.06] - 2013-10-19
- faster sort, 
- dynamic arrays, 
- replaced sorted float array with indirection array

## [0.1.05] - 2013-10-18
- fixed bug in sort; 
- removes default constructor; 
- dynamic memory

## [0.1.04] - 2013-10-17
- added getAverage(uint8_t) - kudo's to Sembazuru

## [0.1.03] - 2013-09-30
- added \_sorted flag, 
- minor refactor

## [0.1.02] - 2012-03-15  
- added ??

## [0.1.01] - 2011-02-22
- added remarks from CodingBadly

## [0.1.00] - 2011-02-16
- initial version



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2021-01-03
// PURPOSE: unit tests for the RunningMedian
//          https://github.com/RobTillaart/RunningMedian
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)


#include <ArduinoUnitTests.h>


#include "Arduino.h"
#include "RunningMedian.h"



unittest_setup()
{
}

unittest_teardown()
{
}


unittest(test_constructor)
{
  fprintf(stderr, "VERSION: %s\n", RUNNING_MEDIAN_VERSION);

  RunningMedian samples = RunningMedian(5);
  assertEqual(5, samples.getSize());
  assertEqual(0, samples.getCount());

  // TODO default values?
}


unittest(test_basic_add)
{
  fprintf(stderr, "VERSION: %s\n", RUNNING_MEDIAN_VERSION);

  RunningMedian samples = RunningMedian(5);

  int cnt = 0;
  for (int i = 0, cnt = 0; i < 50; i+=10)
  {
    samples.add(i);
    cnt++;
    assertEqual(cnt, samples.getCount());
  }
  assertEqual(5, samples.getSize());
  assertEqualFloat(20, samples.getMedian(), 0.0001);
  assertEqualFloat(20, samples.getAverage(), 0.0001);
  assertEqualFloat(00, samples.getLowest(), 0.0001);
  assertEqualFloat(40, samples.getHighest(), 0.0001);
  
  samples.add(100);  // 6th element
  assertEqual(5, samples.getSize());
  assertEqual(5, samples.getCount());

  assertEqualFloat(30, samples.getMedian(), 0.0001);
  assertEqualFloat(40, samples.getAverage(), 0.0001);
  assertEqualFloat(10, samples.getLowest(), 0.0001);
  assertEqualFloat(100, samples.getHighest(), 0.0001);
  
  samples.clear();
  assertEqual(5, samples.getSize());
  assertEqual(0, samples.getCount());
}


unittest(test_big)
{
  fprintf(stderr, "VERSION: %s\n", RUNNING_MEDIAN_VERSION);

  RunningMedian samples = RunningMedian(100);
  assertEqual(100, samples.getSize());
  assertEqual(0, samples.getCount());

  for (int i = 0; i < 110; i++)
  {
    samples.add(i);
  }
  assertEqual(100, samples.getSize());
  assertEqual(100, samples.getCount());

  assertEqualFloat(59.5, samples.getMedian(), 0.0001);
  assertEqualFloat(59.5, samples.getAverage(), 0.0001);
  assertEqualFloat(10, samples.getLowest(), 0.0001);
  assertEqualFloat(109, samples.getHighest(), 0.0001);

  samples.clear();
  assertEqual(100, samples.getSize());
  assertEqual(0, samples.getCount());
}


unittest_main()

// --------


RunningMedian.cpp
//
//    FILE: RunningMedian.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.3.8
// PURPOSE: RunningMedian library for Arduino
//
//  HISTORY: see changelog.md


#include "RunningMedian.h"


RunningMedian::RunningMedian(const uint8_t size)
{
  _size = size;
  if (_size < MEDIAN_MIN_SIZE) _size = MEDIAN_MIN_SIZE;
#if !RUNNING_MEDIAN_USE_MALLOC
  if (_size > MEDIAN_MAX_SIZE) _size = MEDIAN_MAX_SIZE;
#endif

#if RUNNING_MEDIAN_USE_MALLOC
  _values = (float *) malloc(_size * sizeof(float));
  _sortIdx  = (uint8_t *) malloc(_size * sizeof(uint8_t));
#endif
  clear();
}


RunningMedian::~RunningMedian()
{
  #if RUNNING_MEDIAN_USE_MALLOC
  free(_values);
  free(_sortIdx);
  #endif
}


//  resets all internal counters
void RunningMedian::clear()
{
  _count = 0;
  _index = 0;
  _sorted = false;
  for (uint8_t i = 0; i < _size; i++)
  {
    _sortIdx[i] = i;
  }
}


//  adds a new value to the data-set
//  or overwrites the oldest if full.
void RunningMedian::add(float value)
{
  _values[_index++] = value;
  if (_index >= _size) _index = 0; //  wrap around
  if (_count < _size) _count++;
  _sorted = false;
}


float RunningMedian::getMedian()
{
  if (_count == 0) return NAN;

  if (_sorted == false) sort();

  if (_count & 0x01)  //  is it odd sized?
  {
    return _values[_sortIdx[_count / 2]];
  }
  return (_values[_sortIdx[_count / 2]] + _values[_sortIdx[_count / 2 - 1]]) / 2;
}


float RunningMedian::getQuantile(float quantile)
{
  if (_count == 0) return NAN;

  if ((quantile < 0) || (quantile > 1)) return NAN;

  if (_sorted == false) sort();

  const float index = (_count - 1) * quantile;
  const uint8_t lo  = floor(index);
  const uint8_t hi  = ceil(index);
  const float qs    = _values[_sortIdx[lo]];
  const float h     = (index - lo);

  return (1.0 - h) * qs + h * _values[_sortIdx[hi]];
}


float RunningMedian::getAverage()
{
  if (_count == 0) return NAN;

  float sum = 0;
  for (uint8_t i = 0; i < _count; i++)
  {
    sum += _values[i];
  }
  return sum / _count;
}


//  nMedians is the spread, or the middle N
float RunningMedian::getAverage(uint8_t nMedians)
{
  //  handle special cases.
  if ((_count == 0) || (nMedians == 0)) return NAN;
  if (_count == 1) return _values[0];
  if (_count == 2) return (_values[0] + _values[1]) * 0.5;

  //  nMedians can not be larger than current nr of elements.
  if (_count <= nMedians) return getAverage();

  //  _count is at least 3 from here

  //  Eliminate the bias when the nMedians would fall slightly
  //  to the left or right of the centre.
  //  If count and nMedians are not both odd or both even reduce
  //  the spread by 1 to make them the same.
  //  If nMedians becomes 0 correct this. to 2.
  if ((_count & 0x01) != (nMedians & 0x01))
  {
     --nMedians;
     //  nmedians can not become 0
     if (nMedians == 0) nMedians = 2;
  }

  uint8_t start = (_count - nMedians) / 2;
  uint8_t stop = start + nMedians;

  if (_sorted == false) sort();

  float sum = 0;
  for (uint8_t i = start; i < stop; i++)
  {
    sum += _values[_sortIdx[i]];
  }
  return sum / nMedians;
}


float RunningMedian::getElement(const uint8_t n)
{
  if ((_count == 0) || (n >= _count)) return NAN;

  uint8_t pos = _index + n;
  if (pos >= _count)  //  faster than %
  {
    pos -= _count;
  }
  return _values[pos];
}


float RunningMedian::getSortedElement(const uint8_t n)
{
  if ((_count == 0) || (n >= _count)) return NAN;

  if (_sorted == false) sort();
  return _values[_sortIdx[n]];
}


//  n can be max <= half the (filled) size
float RunningMedian::predict(const uint8_t n)
{
  uint8_t mid = _count / 2;
  if ((_count == 0) || (n >= mid)) return NAN;

  float med = getMedian();  //  takes care of sorting !
  if (_count & 0x01)        //  odd # elements
  {
    return max(med - _values[_sortIdx[mid - n]], _values[_sortIdx[mid + n]] - med);
  }
  //  even # elements
  float f1 = (_values[_sortIdx[mid - n]] + _values[_sortIdx[mid - n - 1]]) / 2;
  float f2 = (_values[_sortIdx[mid + n]] + _values[_sortIdx[mid + n - 1]]) / 2;
  return max(med - f1, f2 - med) / 2;
}


void RunningMedian::setSearchMode(uint8_t searchMode)
{
  if (searchMode == 1) _searchMode = 1;
  else _searchMode = 0;
}


uint8_t RunningMedian::getSearchMode()
{
  return _searchMode;
}


////////////////////////////////////////////////////////////
//
//  PRIVATE
//

//  insertion sort - _searchMode = linear or binary.

void RunningMedian::sort()
{
  uint16_t lo   = 0;
  uint16_t hi   = 0;
  uint16_t mi   = 0;
  uint16_t temp = 0;

  for (uint16_t i = 1; i < _count; i++)
  {
    temp = _sortIdx[i];
    float f = _values[temp];

    //  handle special case f is smaller than all elements first.
    //  only one compare needed, improves linear search too.
    if (f <= _values[_sortIdx[0]])
    {
      hi = 0;
    }
    else
    {
      if (_searchMode == 0)
      {
        hi = i;
        //  find insertion point with linear search
        while ((hi > 0) && (f < _values[_sortIdx[hi - 1]]))
        {
          hi--;
        }
      }
      else if (_searchMode == 1)
      {
        //  find insertion point with binary search
        lo = 0;
        hi = i;
        //  be aware there might be duplicates
        while (hi - lo > 1)
        {
          mi = (lo + hi) / 2;
          if (f < _values[_sortIdx[mi]])
          {
            hi = mi;
          }
          else
          {
            lo = mi;
          }
        }
      }
    }

    //  move elements to make space
    uint16_t k = i;
    while (k > hi)
    {
      _sortIdx[k] = _sortIdx[k - 1];
      k--;
    }

    //  insert at right spot.
    _sortIdx[k] = temp;

  }
  _sorted = true;

  // //  verify sorted
  // for (int i = 0; i < _count; i++)
  // {
    // if (i%5 == 0) Serial.println();
    // Serial.print("\t");
    // Serial.print(_values[_sortIdx[i]]);
  // }
  // Serial.println("\n");
}



/*
 split version of pre-0.3.7 sort - bit faster

void RunningMedian::sort()
{
  // insertSort
  for (uint16_t i = 1; i < _count; i++)
  {
    uint16_t hi = i;
    uint16_t temp = _sortIdx[hi];
    float f = _values[temp];
    while ((hi > 0) && (f < _values[_sortIdx[hi - 1]]))
    {
      hi--;
    }

    //  move elements to make space
    uint16_t k = i;
    while (k > hi)
    {
      _sortIdx[k] = _sortIdx[k - 1];
      k--;
    }

    //  insert at right spot.
    _sortIdx[k] = temp;
  }
  _sorted = true;
  //  //  verify sorted
  // for (int i = 0; i < _count; i++)
  // {
    // if (i%5 == 0) Serial.println();
    // Serial.print("\t");
    // Serial.print(_values[_sortIdx[i]]);
  // }
  // Serial.println("\n");
}
*/


/*
//  straightforward insertion sort - PRE-0.3.7
void RunningMedian::sort()
{
  for (uint16_t i = 1; i < _count; i++)
  {
    uint16_t z = i;
    uint16_t temp = _sortIdx[z];
    while ((z > 0) && (_values[temp] < _values[_sortIdx[z - 1]]))
    {
      _sortIdx[z] = _sortIdx[z - 1];
      z--;
    }
    _sortIdx[z] = temp;
  }
  _sorted = true;

  //  //  verify sorted
  // for (int i = 0; i < _count; i++)
  // {
    // if (i%5 == 0) Serial.println();
    // Serial.print("\t");
    // Serial.print(_values[_sortIdx[i]]);
  // }
  // Serial.println("\n");
}
*/

// -- END OF FILE --



RunningMedian.h
#pragma once
//
//    FILE: RunningMedian.h
//  AUTHOR: Rob Tillaart
// PURPOSE: RunningMedian library for Arduino
// VERSION: 0.3.8
//     URL: https://github.com/RobTillaart/RunningMedian
//     URL: http://arduino.cc/playground/Main/RunningMedian
// HISTORY: See RunningMedian.cpp
//


#include "Arduino.h"

#define RUNNING_MEDIAN_VERSION        (F("0.3.8"))


//  fall back to fixed storage for dynamic version => remove true
#ifndef RUNNING_MEDIAN_USE_MALLOC
#define RUNNING_MEDIAN_USE_MALLOC     true
#endif


//  MEDIAN_MIN_SIZE  should at least be 3 to be practical,
#ifndef MEDIAN_MIN_SIZE
#define MEDIAN_MIN_SIZE               3
#endif


#ifndef MEDIAN_MAX_SIZE
#ifdef RUNNING_MEDIAN_USE_MALLOC
//  max 250 to not overflow uint8_t internal variables
#define MEDIAN_MAX_SIZE               255
#else
//  using fixed memory will be limited to 19 elements.
#define MEDIAN_MAX_SIZE               19
#endif
#endif


class RunningMedian
{
public:
  //  # elements in the internal buffer
  //  odd sizes results in a 'real' middle element and will be a bit faster.
  //  even sizes takes the average of the two middle elements as median
  explicit RunningMedian(const uint8_t size);
  ~RunningMedian();

  //  resets internal buffer and variables
  void    clear();
  //  adds a new value to internal buffer, optionally replacing the oldest element.
  void    add(const float value);
  //  returns the median == middle element
  float   getMedian();

  //  returns the Quantile
  float   getQuantile(const float quantile);

  //  returns average of the values in the internal buffer
  float   getAverage();
  //  returns average of the middle nMedian values, removes noise from outliers
  float   getAverage(uint8_t nMedian);

  float   getHighest() { return getSortedElement(_count - 1); };
  float   getLowest()  { return getSortedElement(0); };

  //  get n-th element from the values in time order
  float   getElement(const uint8_t n);
  //  get n-th element from the values in size order
  float   getSortedElement(const uint8_t n);
  //  predict the max change of median after n additions
  float   predict(const uint8_t n);

  uint8_t getSize()    { return _size; };
  //  returns current used elements, getCount() <= getSize()
  uint8_t getCount()   { return _count; };
  bool    isFull()     { return (_count == _size); }

  //  EXPERIMENTAL  (might change in the future)
  //  searchMode defines how the internal insertionSort works
  //  can be used to optimize performance.
  //  0 = LINEAR_SEARCH   1 = BINARY_SEARCH
  void    setSearchMode(uint8_t searchMode = 0);
  uint8_t getSearchMode();


protected:
  boolean   _sorted;    //  _sortIdx{} is up to date
  uint8_t   _size;      //  max number of values
  uint8_t   _count;     //  current number of values <= size
  uint8_t   _index;     //  next index to add


  //  _values holds the elements themself
  //  _sortIdx holds the index for sorted
#if RUNNING_MEDIAN_USE_MALLOC
  float *   _values;
  uint8_t * _sortIdx;
#else
  float   _values[MEDIAN_MAX_SIZE];
  uint8_t _sortIdx[MEDIAN_MAX_SIZE];
#endif
  void      sort();
  uint8_t   _searchMode = 0;
};


// -- END OF FILE --

